src/app/clusters/laundry-washer-controls-server/laundry-washer-controls-server.cpp

```diff
@@ -0,0 +1,231 @@
+    return (err == CHIP_NO_ERROR) ? (EMBER_ZCL_STATUS_SUCCESS) : (EMBER_ZCL_STATUS_FAILURE);
+}
+
+EmberAfStatus LaundryWasherControlsServer::GetSpinSpeedList(EndpointId endpointId, SpinSpeedListCharSpan ** spinSpeedList,
+                                                            size_t & size)
+{
+    CHIP_ERROR err = mLaundryWasherDataProvider.LoadSpinSpeedList(endpointId, LaundryWasherControls::Id, spinSpeedList, size);
+    return (err == CHIP_NO_ERROR) ? (EMBER_ZCL_STATUS_SUCCESS) : (EMBER_ZCL_STATUS_FAILURE);
+}
+
+EmberAfStatus LaundryWasherControlsServer::SetSupportedRinsesList(EndpointId endpointId,
+                                                                  const SupportedRinsesList & supportedRinsesList)
+{
+    CHIP_ERROR err =
+        mLaundryWasherDataProvider.StoreSupportedRinsesList(endpointId, LaundryWasherControls::Id, supportedRinsesList);
+    return (err == CHIP_NO_ERROR) ? (EMBER_ZCL_STATUS_SUCCESS) : (EMBER_ZCL_STATUS_FAILURE);
+}
+
+EmberAfStatus LaundryWasherControlsServer::GetSupportedRinsesList(EndpointId endpointId,
+                                                                  SupportedRinsesListSpan ** supportedRinsesList, size_t & size)
+{
+    CHIP_ERROR err =
+        mLaundryWasherDataProvider.LoadSupportedRinsesList(endpointId, LaundryWasherControls::Id, supportedRinsesList, size);
+    return (err == CHIP_NO_ERROR) ? (EMBER_ZCL_STATUS_SUCCESS) : (EMBER_ZCL_STATUS_FAILURE);
+}
+
+void LaundryWasherControlsServer::ReleaseSpinSpeedList(SpinSpeedListCharSpan * spinSpeedList)
+{
+    mLaundryWasherDataProvider.ReleaseSpinSpeedList(spinSpeedList);
+}
+
+EmberAfStatus LaundryWasherControlsServer::ClearSpinSpeedList(EndpointId endpointId)
+{
+    CHIP_ERROR err = mLaundryWasherDataProvider.ClearSpinSpeedList(endpointId, LaundryWasherControls::Id);
+    return (err == CHIP_NO_ERROR) ? (EMBER_ZCL_STATUS_SUCCESS) : (EMBER_ZCL_STATUS_FAILURE);
+}
+
+void LaundryWasherControlsServer::ReleaseSupportedRinsesList(SupportedRinsesListSpan * supportedRinsesList)
+{
+    mLaundryWasherDataProvider.ReleaseSupportedRinsesList(supportedRinsesList);
+}
+
+EmberAfStatus LaundryWasherControlsServer::ClearSupportedRinsesList(EndpointId endpointId)
+{
+    CHIP_ERROR err = mLaundryWasherDataProvider.ClearSupportedRinsesList(endpointId, LaundryWasherControls::Id);
+    return (err == CHIP_NO_ERROR) ? (EMBER_ZCL_STATUS_SUCCESS) : (EMBER_ZCL_STATUS_FAILURE);
+}
+
+/**********************************************************
+ * LaundryWasherControlsServer private methods
+ *********************************************************/
+CHIP_ERROR LaundryWasherControlsServer::ReadSpinSpeeds(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder)
+{
+    SpinSpeedListCharSpan * spinSpeedList = nullptr;
+    size_t size                           = 0;
+    EndpointId endpointId                 = aPath.mEndpointId;
+    GetSpinSpeedList(endpointId, &spinSpeedList, size);
```

Comment:
Again: this is forcing the whole list into memory at once.  Why are we doing this?

I feel like we are going in circles here.  Perhaps take a look at the delegate class in https://github.com/project-chip/connectedhomeip/pull/27504 for the sort of API shape that we are doing elsewhere?

Reply:
It's changed